File: examples/cpp/run_sync_model_bufcount/run_sync_model_bufcount.cpp
/*
 * Copyright (C) 2018- DEEPX Ltd.
 * All rights reserved.
 *
 * This software is the property of DEEPX and is provided exclusively to customers
 * who are supplied with DEEPX NPU (Neural Processing Unit).
 * Unauthorized sharing or usage is strictly prohibited by law.
 */

#include "dxrt/dxrt_api.h"
#include "dxrt/extern/cxxopts.hpp"
#include "../include/logger.h"
#include <string>
#include <iostream>


int main(int argc, char* argv[])
{
    std::string model_path;
    int loop_count;
    bool verbose;

    auto &log = dxrt::Logger::GetInstance();

    cxxopts::Options options("run_sync_model", "Run synchronous model inference");
    options.add_options()
        ("m,model", "Path to model file (.dxnn)", cxxopts::value<std::string>(model_path))
        ("l,loops", "Number of inference loops", cxxopts::value<int>(loop_count)->default_value("1"))
        ("v,verbose", "Enable verbose/debug logging", cxxopts::value<bool>(verbose)->default_value("false"))
        ("h,help", "Print usage");

    try
    {
        auto result = options.parse(argc, argv);

        if (result.count("help") || !result.count("model"))
        {
            std::cout << options.help() << std::endl;
            return result.count("help") ? 0 : -1;
        }

        if (verbose) {
            log.SetLevel(dxrt::Logger::Level::LOGLEVEL_DEBUG);
        }
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("Error parsing arguments: ") + e.what());
        std::cout << options.help() << std::endl;
        return -1;
    }

    log.Info("Start run_sync_model test for model: " + model_path);

    try
    {

        // create inference engine instance with model
        dxrt::InferenceOption option_1;
        option_1.bufferCount = 6; // set buffer count to 6
        log.Info("Creating InferenceEngine_1 with buffer count: " + std::to_string(option_1.bufferCount));
        dxrt::InferenceEngine ie_1(model_path, option_1);

        dxrt::InferenceOption option_2;
        option_2.bufferCount = 3; // set buffer count to 3
        log.Info("Creating InferenceEngine_2 with buffer count: " + std::to_string(option_2.bufferCount));
        dxrt::InferenceEngine ie_2(model_path, option_2);

        // create temporary input buffer for example
        std::vector<uint8_t> inputPtr_1(ie_1.GetInputSize(), 0);
        std::vector<uint8_t> inputPtr_2(ie_2.GetInputSize(), 0);

        auto start = std::chrono::high_resolution_clock::now();

        // inference loop
        for(int i = 0; i < loop_count; ++i)
        {

            // inference synchronously, use one npu core
            auto outputs_1 = ie_1.Run(inputPtr_1.data());
            auto outputs_2 = ie_2.Run(inputPtr_2.data());

            log.Debug("Inference outputs (" + std::to_string(i) + ")");

            // now there is no post processing
            (void)outputs_1;
            (void)outputs_2;

        }

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> duration = end - start;

        double total_time = duration.count();
        double avg_latency = total_time / static_cast<double>(loop_count);
        double fps = 1000.0 / avg_latency;

        log.Info("-----------------------------------");
        log.Info("Total Time: " + std::to_string(total_time) + " ms");
        log.Info("Average Latency: " + std::to_string(avg_latency) + " ms");
        log.Info("FPS: " + std::to_string(fps) + " frame/sec");
        log.Info("Success");
        log.Info("-----------------------------------");
    }
    catch (const dxrt::Exception& e)
    {
        log.Error(std::string("dxrt::Exception: ") + e.what());
        return -1;
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("std::exception: ") + e.what());
        return -1;
    }
    catch(...)
    {
        log.Error("Exception");
        return -1;
    }

    return 0;
}

File: examples/python/run_sync_model_bufcount.py
#
# Copyright (C) 2018- DEEPX Ltd.
# All rights reserved.
#
# This software is the property of DEEPX and is provided exclusively to customers 
# who are supplied with DEEPX NPU (Neural Processing Unit). 
# Unauthorized sharing or usage is strictly prohibited by law.
#

import numpy as np
import os
import argparse
import time
from dx_engine import InferenceEngine, InferenceOption
from logger import Logger, LogLevel


def parse_args():
    parser = argparse.ArgumentParser(description="Run synchronous model inference")
    parser.add_argument("--model", "-m", type=str, required=True, help="Path to model file (.dxnn)")
    parser.add_argument("--loops", "-l", type=int, default=1, help="Number of inference loops (default: 1)")
    parser.add_argument("--verbose", "-v", action="store_true", default=False, help="Enable debug logging")
    args = parser.parse_args()

    if not os.path.exists(args.model):
        parser.error(f"Model path '{args.model}' does not exist.")
    
    if args.verbose:
        logger = Logger()
        logger.set_level(LogLevel.DEBUG)
    
    return args


if __name__ == "__main__":
    args = parse_args()
    logger = Logger()
        
    logger.info(f"Start run_sync_model test for model: {args.model}")
    
    try:
        # create inference engine instance with model
        ie_1_option = InferenceOption()
        ie_1_option.buffer_count = 6
        logger.info(f"Buffer count for IE 1: {ie_1_option.buffer_count}")
        ie_2_option = InferenceOption()
        ie_2_option.buffer_count = 3
        logger.info(f"Buffer count for IE 2: {ie_2_option.buffer_count}")
        with InferenceEngine(args.model, ie_1_option) as ie_1, InferenceEngine(args.model, ie_2_option) as ie_2:

            # NOTE: np.zeros() uses COW zero pages — all virtual pages share one
            # physical page. PCIe DMA driver's get_user_pages() then sees duplicate
            # physical pages in the SG list and fails with EFAULT.
            # np.empty() + explicit fill forces unique physical page allocation.
            _buf_1 = np.empty(ie_1.get_input_size(), dtype=np.uint8)
            _buf_1.fill(0)
            input_1 = [_buf_1]
            _buf_2 = np.empty(ie_2.get_input_size(), dtype=np.uint8)
            _buf_2.fill(0)
            input_2 = [_buf_2]

            start = time.perf_counter()
            # inference loop
            for i in range(args.loops):

                # inference synchronously 
                # use only one npu core 
                # ownership of the outputs is transferred to the user 
                outputs_1 = ie_1.run(input_1)
                outputs_2 = ie_2.run(input_2)

                # post processing 
                #postProcessing(outputs)
                logger.debug(f"Inference outputs {i}")
                
            end = time.perf_counter()
            total_time_ms = (end -start) * 1000
            avg_latency = total_time_ms / args.loops
            fps = 1000.0/ avg_latency if avg_latency > 0 else 0.0
            
            logger.info("-----------------------------------")
            logger.info(f"Total Time: {total_time_ms:.3f} ms")
            logger.info(f"Average Latency: {avg_latency:.3f} ms")
            logger.info(f"FPS: {fps:.2f} frame/sec")
            logger.info("Success")
            logger.info("-----------------------------------")
                
    except Exception as e:
        logger.error(f"Exception: {str(e)}")
        exit(-1)
        
    exit(0)